### PR TITLE
Use ESM

### PR DESCRIPTION
**Pull request by jablko**
I think this is a prerequisite of upgrading to [hast-util-to-estree v2](https://github.com/syntax-tree/hast-util-to-estree/commit/afc22d234775977f2e529baca1f344b3fb9f71c0)? Otherwise I get the following error, because hast-util-to-estree is pure ESM, which `require()` doesn't support:
```
 FAIL  test/index.test.js
  ● Test suite failed to run

    Jest encountered an unexpected token

    This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.

    By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/en/ecmascript-modules for how to enable it.
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/en/configuration.html

    Details:

    /home/nottheoilrig/mdx/node_modules/hast-util-to-estree/index.js:58
    import {stringify as commas} from 'comma-separated-tokens'
    ^^^^^^

    SyntaxError: Cannot use import statement outside a module

    > 1 | const toEstree = require('hast-util-to-estree')
        |                  ^
      2 | const walk = require('estree-walker').walk
      3 | const periscopic = require('periscopic')
      4 | const estreeToJs = require('./estree-to-js')

      at Runtime.createScriptFromCode (../../node_modules/jest-runtime/build/index.js:1350:14)
      at Object.<anonymous> (mdx-hast-to-jsx.js:1:18)
```

To replace `require()` with `import` I ran the following, plus manual fix-ups:
```Shell
$ sed --in-place '
  s/\(const\|var\) \(.*\) = require(\(.*\))/import \2 from \3/
  s/module.exports = require(\(.*\))/export * from \1/
  s/module.exports = /export default /
' $(
    git ls-files \
      docs \
      packages/loader \
      packages/mdx \
      packages/parcel-plugin-mdx \
      packages/vue-loader \
  )
```

Depends on https://github.com/facebook/jest/pull/11790
Depends on https://github.com/webpack/webpack/pull/14077

### CODE REVIEW

File: .eslintrc.yml [hunk 1]
```diff
@@ -83,9 +83,9 @@ overrides:
       - '*.tsx'
     extends: 'plugin:@typescript-eslint/recommended'
     settings:
-      import:
-        resolver:
-          typescript: null
+      import/resolver:
+        typescript:
+          project: packages/*/types
```
Comment (jablko) [inline .eslintrc.yml:88]: So [eslint-import-resolver-typescript](https://github.com/alexgorbatchev/eslint-import-resolver-typescript#configuration) can find `@mdx-js/mdx.Options`:
```
  ✘  https://google.com/#q=import%2Fnamed

     Options not found in '@mdx-js/mdx'

     packages/runtime/types/index.d.ts:4:9
     2 | 
     3 | import {FunctionComponent} from 'react'
   > 4 | import {Options} from '@mdx-js/mdx'
       |         ^
     5 | import {ComponentsProp} from '@mdx-js/react'
     6 | 
     7 | /**

File: packages/loader/package.json [hunk 2]
```diff
@@ -35,8 +35,8 @@
     "loader"
   ],
   "scripts": {
-    "test-api": "jest test",
-    "test-coverage": "jest test --coverage",
+    "test-api": "NODE_OPTIONS=--experimental-vm-modules jest test",
+    "test-coverage": "NODE_OPTIONS=--experimental-vm-modules jest test --coverage",
```
Comment (jablko) [inline packages/loader/package.json:39]: `--experimental-vm-modules` for [tests written in ESM](https://jestjs.io/docs/ecmascript-modules#:~:text=execute%20node%20with%20--experimental-vm-modules).

File: packages/loader/package.json [hunk 3]
```diff
@@ -49,11 +49,11 @@
     "loader-utils": "^2.0.0"
   },
   "devDependencies": {
-    "memory-fs": "^0.5.0",
     "react": "^17.0.0",
     "react-dom": "^17.0.0",
     "remark-slug": "^6.0.0",
-    "webpack": "^4.0.0"
+    "webpack": "^5.51.1"
```
Comment (jablko) [inline packages/loader/package.json:55]: Upgrade to webpack 5 for https://github.com/webpack/loader-runner/pull/40, although it's still missing https://github.com/webpack/webpack/pull/14077.

File: packages/loader/test/index.test.js [hunk 4]
```diff
@@ -1,19 +1,16 @@
...
+import React from 'react'
+import {renderToString} from 'react-dom/server'
+import remarkSlug from 'remark-slug'
 
 const transform = (filePath, options) => {
   return new Promise((resolve, reject) => {
-    // Webpack 5: const fs = new MemoryFs()
     const compiler = webpack({
-      context: __dirname,
+      context: fileURLToPath(new URL('.', import.meta.url)),
```
Comment (jablko) [inline packages/loader/test/index.test.js:13]: [No `__filename` or `__dirname` in ESM](https://nodejs.org/api/esm.html#esm_no_filename_or_dirname).

File: packages/loader/test/index.test.js [hunk 5]
```diff
@@ -32,55 +29,41 @@ const transform = (filePath, options) => {
...
-        // resolve(
-        //   {source: fs.readFileSync(path.join(__dirname, '..', 'dist', 'main.js'), 'utf8')}
-        // )
-        resolve(stats.toJson().modules.find(m => m.name === filePath))
+        resolve({
+          source: fs.readFileSync(
+            new URL('../dist/main.js', import.meta.url),
+            'utf8'
+          )
+        })
```
Comment (jablko) [inline packages/loader/test/index.test.js:48]: Upgrade to webpack 5 with the help of the existing comments.

File: packages/mdx/types/index.d.ts [hunk 6]
```diff
@@ -2,66 +2,39 @@
 
 import {Plugin, Compiler, Processor} from 'unified'
 
-declare namespace mdx {
```
Comment (jablko) [inline packages/mdx/types/index.d.ts:5]: Update types correspondingly.

File: packages/parcel-plugin-mdx/src/index.js [hunk 7]
```diff
@@ -1,3 +1,8 @@
-module.exports = function (bundler) {
-  bundler.addAssetType('mdx', require.resolve('./MDXAsset.js'))
+import {createRequire} from 'module'
+
+export default function (bundler) {
+  bundler.addAssetType(
+    'mdx',
+    createRequire(import.meta.url).resolve('./MDXAsset.js')
```
Comment (jablko) [inline packages/parcel-plugin-mdx/src/index.js:6]: `createRequire()` works, until Jest supports [`import.meta.resolve()`](https://github.com/facebook/jest/issues/9505#issue-559108665).

File: packages/react/package.json [hunk 8]
```diff
@@ -19,11 +19,11 @@
     "Christian Murphy <christian.murphy.42@gmail.com>"
   ],
   "license": "MIT",
-  "main": "dist/mdx-react.js",
+  "main": "dist/mdx-react.cjs",
```
Comment (jablko) [inline packages/react/package.json:22]: `"type": "module"` implies that `dist/mdx-react.js` is ESM. `dist/mdx-react.cjs` is identical but explicitly CJS. This avoids:
```
FAIL test/index.test.js
  ● Test suite failed to run

    SyntaxError: The requested module '@mdx-js/react' does not provide an export named 'MDXProvider'

      at Runtime.linkAndEvaluateModule (../../node_modules/jest-runtime/build/index.js:669:5)
      at TestScheduler.scheduleTests (../../node_modules/@jest/core/build/TestScheduler.js:333:13)
      at runJest (../../node_modules/@jest/core/build/runJest.js:387:19)
      at _run10000 (../../node_modules/@jest/core/build/cli/index.js:408:7)
      at runCLI (../../node_modules/@jest/core/build/cli/index.js:261:3)

File: packages/vue-loader/test/test.js [hunk 9]
```diff
@@ -1,15 +1,13 @@
...
-const {mount} = require('@vue/test-utils')
-const vueMergeProps = require('babel-helper-vue-jsx-merge-props')
-const {mdx} = require('../../vue')
+import {fileURLToPath} from 'url'
+import fs from 'fs'
+import webpack from 'webpack'
+import {mount} from '@vue/test-utils'
+import remarkSlug from 'remark-slug'
 
-// See `loader`’s tests for how to upgrade these to webpack 5.
```
Comment (jablko) [inline packages/vue-loader/test/test.js:8]: Upgrade to webpack 5, same as `packages/loader`.

**Comment by jablko** [comment]
FYI, [current CI failure](https://github.com/mdx-js/mdx/runs/3451398035?check_suite_focus=true#step:5:322) is pending https://github.com/facebook/jest/pull/11790.

**Comment by ChristianMurphy** [comment]
Thanks @jablko!
ESM support is of interest and a goal for the project.

A related project, XDM, has ESM support https://github.com/wooorm/xdm you may want to check it out.
It also includes workarounds for WebPack's lack of ESM loader support https://github.com/wooorm/xdm#webpack (though ideally ESM should be natively supported upstream), and native TypeScript support.

There's been discussion of MDX and XDM merging back together, which would likely bring ESM support into MDX.
Which does leave things in a bit of a lurch.
This PR and the potential merge with XDM are both large changes which would significantly overlap. :thinking: 
It may be worth waiting a bit to see where that goes.

In the meantime, could you expand a bit on what features/fixes from hast-util-to-estree v2 you are looking for?
May be able to offer some pointers while  MDX and XDM get sorted out.

**Comment by wooorm** [comment]
Thanks for your work on this Jack.
You may have seen that I since cleaned the repo, ported xdm back into it, and am hard at work working on the new website (https://v2.mdxjs.com). Part of that effort was also ESM, hence I’ll close this.
While the website is not done, and the code has not been released, I would appreciate to hear your thoughts on them if you have any!

**Pull request closed without merging**